### PR TITLE
Task Cancellation

### PR DESCRIPTION
**Pull request by JosephEarl**
Implement task cancellation as per [Task Cancellation](https://msdn.microsoft.com/en-us/library/dd997396(v=vs.110).aspx).

* Add CancellationToken and CancellationTokenSource to match [system.threading.cancellationtoken](https://msdn.microsoft.com/en-us/library/system.threading.cancellationtoken(v=vs.110).aspx) and [system.threading.cancellationtokensource](https://msdn.microsoft.com/en-us/library/system.threading.cancellationtokensource(v=vs.110).aspx)
* Add methods to Task that accept CancellationTokens
* Add AppLinkResolver method that accepts CancellationTokens

### CODE REVIEW

File: Bolts/src/bolts/Task.java [hunk 1]
```diff
@@ -485,6 +610,7 @@ public Void then(Task<TResult> task) {
     }, executor);
   }
 
+
```
Comment (JosephEarl) [inline Bolts/src/bolts/Task.java]: extra newline

File: BoltsTest/src/bolts/TaskTest.java [hunk 2]
```diff
@@ -156,6 +170,69 @@ public Void then(Task<Integer> task) {
...
+    });
+
+    while (true) {
+      synchronized (cancelLock) {
+        if (waitingToBeCancelled.get()) {
+          cts.cancel();
+          cancelLock.notify();
+          break;
+        }
+        try {
```
Comment (JosephEarl) [inline BoltsTest/src/bolts/TaskTest.java]: Move out of synchronized block

File: Bolts/src/bolts/Task.java [hunk 3]
```diff
@@ -365,6 +413,7 @@ public Void then(Task<Object> task) {
     return makeVoid().continueWithTask(predicateContinuation.get(), executor);
   }
 
+
```
Comment (JosephEarl) [inline Bolts/src/bolts/Task.java]: Extra newline

File: Bolts/src/bolts/Task.java [hunk 4]
```diff
@@ -219,7 +237,14 @@ public void run() {
    * from the callable.
    */
   public static <TResult> Task<TResult> call(final Callable<TResult> callable) {
-    return call(callable, IMMEDIATE_EXECUTOR);
+    return call(callable, (CancellationToken) null);
```
Comment (JosephEarl) [inline Bolts/src/bolts/Task.java]: Use 3-arg call with IMMEDIATE_EXECUTOR  to avoid cast

File: Bolts/src/bolts/Task.java [hunk 5]
```diff
@@ -348,13 +373,36 @@ public Void then(Task<Object> task) {
    * Continues a task with the equivalent of a Task-based while loop, where the body of the loop is
```
Comment (grantland) [inline Bolts/src/bolts/Task.java]: line 369: use 4 param

File: Bolts/src/bolts/Task.java [hunk 6]
```diff
@@ -186,7 +185,14 @@ public void waitForCompletion() throws InterruptedException {
    * from the callable.
    */
   public static <TResult> Task<TResult> callInBackground(Callable<TResult> callable) {
-    return call(callable, BACKGROUND_EXECUTOR);
+    return callInBackground(callable, null);
```
Comment (grantland) [inline Bolts/src/bolts/Task.java]: Use 3-arg call with `IMMEDIATE_EXECUTOR` to lessen call depth

File: Bolts/src/bolts/Task.java (hunk 5 above)
Comment (grantland) [inline Bolts/src/bolts/Task.java]: line 369: use 4 param

File: Bolts/src/bolts/Task.java [hunk 7]
```diff
@@ -402,11 +462,30 @@ public Void then(Task<TResult> task) {
   }
```
Comment (grantland) [inline Bolts/src/bolts/Task.java]: use 3 param

File: Bolts/src/bolts/Task.java [hunk 8]
```diff
@@ -437,14 +516,37 @@ public Void then(Task<TResult> task) {
...
```
Comment (grantland) [inline Bolts/src/bolts/Task.java]: use 3 param

File: Bolts/src/bolts/Task.java [hunk 9]
```diff
@@ -469,11 +571,34 @@ public Void then(Task<TResult> task) {
    * Runs a continuation when a task completes successfully, forwarding along
```
Comment (grantland) [inline Bolts/src/bolts/Task.java]: Use 3 param

File: Bolts/src/bolts/Task.java [hunk 10]
```diff
@@ -495,6 +621,16 @@ public Void then(Task<TResult> task) {
   }
```
Comment (grantland) [inline Bolts/src/bolts/Task.java]: Use 3 param

File: BoltsTest/src/bolts/CancellationTest.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+import android.test.InstrumentationTestCase;
+
+import java.util.concurrent.CancellationException;
+
+public class CancellationTest extends InstrumentationTestCase {
+  public void testTokenIsCancelled() {
+    CancellationTokenSource cts = new CancellationTokenSource();
+    CancellationToken token = cts.getToken();
+
```
Comment (grantland) [inline BoltsTest/src/bolts/CancellationTest.java]: assertFalse(token.isCancellationRequested());

File: BoltsTest/src/bolts/CancellationTest.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    cts.cancel();
+
+    assertTrue(token.isCancellationRequested());
+  }
+
+  public void testTokenThrowsWhenCancelled() {
+    CancellationTokenSource cts = new CancellationTokenSource();
+    CancellationToken token = cts.getToken();
+
```
Comment (grantland) [inline BoltsTest/src/bolts/CancellationTest.java]: token.throwIfCancellationRequested();

File: BoltsTest/src/bolts/CancellationTest.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    cts.cancel();
+
+    try {
+      token.throwIfCancellationRequested();
+      fail(CancellationException.class.getSimpleName() + " should be thrown");
+    } catch (CancellationException e) {
```
Comment (grantland) [inline BoltsTest/src/bolts/CancellationTest.java]: // do nothing

File: BoltsTest/src/bolts/TaskTest.java [hunk 14]
```diff
@@ -127,6 +128,19 @@ public Integer then(Task<Integer> task) {
     assertTrue(second.isCancelled());
   }
 
+  public void testSynchronousContinuationTokenAlreadyCancelled() {
+    CancellationTokenSource cts = new CancellationTokenSource();
+    cts.cancel();
+    Task<Integer> first = Task.forResult(1);
+    Task<Integer> second = first.continueWith(new Continuation<Integer, Integer>() {
+      public Integer then(Task<Integer> task) {
```
Comment (grantland) [inline BoltsTest/src/bolts/TaskTest.java]: let's verify that continuation block wasn't called

File: README.md [hunk 15]
```diff
@@ -356,6 +356,98 @@ saveAsync(obj1).onSuccessTask(new Continuation<ParseObject, Task<ParseObject>>()
...
+  // Create a new Task
+  Task<Integer>.CompletionSource tcs = Task.create();
+
+  new Thread() {
+    @Override
+    public void run() {
+      // Check if cancelled at start
+      if (ct.isCancellationRequested()) {
+        tcs.setCancelled();
+        ct.throwIfCancellationRequested();
```
Comment (grantland) [inline README.md]: I think this sample code would crash right here due to an uncaught exception in a new thread? We probably want `return;` instead.

File: README.md [hunk 16]
```diff
@@ -356,6 +356,98 @@ saveAsync(obj1).onSuccessTask(new Continuation<ParseObject, Task<ParseObject>>()
...
+      }
+
+      int result = 0;
+      while (result < 100) {
+        // Poll isCancellationRequested in a loop
+        if (ct.isCancellationRequested()) {
+          tcs.setCancelled();
+          ct.throwIfCancellationRequested();
```
Comment (grantland) [inline README.md]: same

File: README.md [hunk 17]
```diff
@@ -356,6 +356,98 @@ saveAsync(obj1).onSuccessTask(new Continuation<ParseObject, Task<ParseObject>>()
...
+```
+
+where `Operation` has some `cancel()` method, you can adapt this to Tasks and CancellationTokens very easily and have a consistent set of asynchronous operations using tasks.
+
+```java
+public Task<String> getStringAsync(CancellationToken ct) {
+  Task<String>.CompletionSource tcs = Task.create();
+  CancellationToken.ListenerId listenerId = new CancellationToken.ListenerId();
+
+  final Operation<String> operation = getStringAsync(new Callback() {
```
Comment (grantland) [inline README.md]: getStringAsyncOperation

**Comment by grantland** [comment]
Let's revert the changes in `AppLinkResolver.java` and `WebViewAppLinkResolver.java`. Adding a method to `AppLinkResolver` is a breaking change since it requires developers to implement an extra method and it doesn't seem like we're utilizing the CT version of the method in `WebViewAppLinkResolver` anyway.

I'm not sure how I feel about `CancellationListener`. I understand you added it to wrap other CancellationToken types from other APIs, but it would be great if we can move it to another PR to discuss.

Otherwise, everything is looking good! You've been making awesome contributions :)

File: Bolts/src/bolts/Task.java (hunk 6 above)
Comment (JosephEarl) [inline Bolts/src/bolts/Task.java]: You mean `BACKGROUND_EXECUTOR`?

**Comment by JosephEarl** [comment]
@grantland updated with suggested changes, will move the `CancellationListener` into another PR

RE `AppLinkResolver` changes - yes, I forgot to also make the changes to `AppLinkNavigation`. 
Perhaps we could create a second `AppLinkResolver2` (or `CancellableAppLinkResolver`) interface and add extra methods to AppLinkNavigation to use that together with CancellationToken.

File: Bolts/src/bolts/CancellationToken.java [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * Propagates notification that operations should be canceled.
+ * <p/>
+ * Create an instance of {@code CancellationTokenSource} and pass the token returned from
+ * {@code #getToken()} to the asynchronous operation(s). Call
+ * {@code CancellationTokenSource#cancel()} to cancel the operations.
+ * <p/>
+ * A {@code CancellationToken} can only be cancelled once - it should not be passed to future operations
+ * once cancelled.
+ */
```
Comment (grantland) [inline Bolts/src/bolts/CancellationToken.java]: `{@link CancellationTokenSource}`

`{@link CancellationTokenSource#getToken()}`

`{@link CancellationTokenSource#cancel()}`

File: Bolts/src/bolts/CancellationToken.java [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+public class CancellationToken {
+
+  private final Object lock = new Object();
+  private boolean cancellationRequested;
+
+  CancellationToken() {
```
Comment (grantland) [inline Bolts/src/bolts/CancellationToken.java]:
```java
/* package */
```

File: Bolts/src/bolts/CancellationToken.java [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+      return cancellationRequested;
+    }
+  }
+
+  /**
+   * Throws an exception if the user has a requested this token be cancelled. May be used to stop
+   * execution of a thread or runnable.
+   *
+   * @throws CancellationException
+   */
```
Comment (grantland) [inline Bolts/src/bolts/CancellationToken.java]:
```java
/**
 * @throws CancellationException if this token has had cancellation requested. 
 * May be used to stop execution of a thread or runnable.
 */
```

File: Bolts/src/bolts/CancellationToken.java [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+  public void throwIfCancellationRequested() throws CancellationException {
+    synchronized (lock) {
+      if (cancellationRequested) {
+        throw new CancellationException();
+      }
+    }
+  }
+
+  boolean tryCancel() {
```
Comment (grantland) [inline Bolts/src/bolts/CancellationToken.java]:
```java
/* package */
```

File: Bolts/src/bolts/CancellationTokenSource.java [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+ */
+package bolts;
+
+import java.util.Locale;
+
+/**
+ * Signals to a CancellationToken that it should be canceled. To create a
+ * {@code CancellationToken} first create a {@code CancellationTokenSource} then call
+ * {@code CancellationTokenSource#getToken()} to retrieve the token for the source.
+ */
```
Comment (grantland) [inline Bolts/src/bolts/CancellationTokenSource.java]: `{@link CancellationToken}`

`{@link #getToken()}`

**Comment by grantland** [comment]
@JosephEarl Thanks! I've added some doc changes and I'm planning on doing another once-over for the API and implementation, so you can wait on updating.

RE AppLinkResolver - I'm not sure AppLinkResolver2 would work since we'd need to accept the base implementation in `AppLinkNavigation#navigateInBackground`, which would also be a breaking change. Are you using AppLinkResolver in your codebase?

**Comment by JosephEarl** [comment]
@grantland I think there would need to be a new `AppLinkNavigation#navigateInBackground` method which accepts a `CancellationToken` anyway, so this would accept `AppLinkResolver2` only. `AppLinkResolver2` would extend `AppLinkResolver`.

**Comment by grantland** [comment]
Re `AppLinkResolver2`: That might be a good solution, but unfortunately I'll have to defer AppLinks discussions to @mingflifb. It'll probably be better in a separate PR.

I also just noticed `CancellationToken.Register(Action)`, which looks be similar to the `CancellationListener` you added. I'm sorry I missed this, but feel free to add that back to this PR if you'd like it to be reviewed again.

Please note that the following isn't about requesting things to be changed, but opening discussion:

TPL defines [`FromCanceled<TResult>(CancellationToken)`](https://msdn.microsoft.com/en-us/library/system.threading.tasks.task.fromcanceled%28v=vs.110%29.aspx). Do we want to support something similar? We already have `<TResult> Task<TResult> cancelled()` and looking at the TPL source code, `FromCanceled` throws if the token isn't actually cancelled, which brings me to believe that it needs to be wrapped with an `if (ct.isCancellationRequested)` anyway. If this is the case, `cancelled()` seems like it should suffice.

Do we want to add `CancellationToken#none()` and `CancellationToken#canBeCanceled()`?

Otherwise, the implementation looks great!

**Comment by JosephEarl** [comment]
OK I'll re-add `CancellationToken.register()`. On issue in comparison with the C# implementation is a need to unregister to avoid leaking memory. Consider something like the following:

    Task<String> getStringAsync(CancellationToken ct) {
        TaskCompletetionSource<String> tcs = new Task.create();
        final Operation<String> operation = getStringOperationAsync(new Callback<String>() {
            void onComplete(String response) {
                tcs.setResult(response);
            }
         }
         ct.register(new CancellationListener() {
             void onCancellationRequested(CancellationToken ct) {
                  operation.cancel();
                  tcs.trySetCancelled();
             }
         });
    }

If the operation completes successfully, the listener is still registered and retaining operation (and the same token may be shared between a number of tasks).

Also the cancellation token may be cancelled some time later after this operation has completed. In this case the code above still attempts to cancel the operation.

To avoid this an `unregister` method is needed. 

And since the `CancellationListener` will be registered (and thus created) after the operation has been created there needs to be some way of identifying the listener when registering/unregistering independent of the listener object itself.

I suppose another option to the ID could be to Capture the cancellation listener so the Capture can be created before the operation. 

`Task.fromCancelled` looks like a sensible addition so I'll add that. What kind of exception do you think is appropriate, `IllegalStateException`? 

Just to be clear on `CancellationToken#none()` and `CancellationToken#canBeCanceled()` - `CancellationToken#none()` returns a token that cannot be cancelled, whereas the token from `CancellationTokenSource` can be cancelled?

**Comment by grantland** [comment]
Re `CanBeCanceled`/`None`:
to be honest, I'm not very experienced with these APIs, but from what I can tell from docs/source you're correct. it also doesn't seem to be 100% necessary atm, so feel free to pass on that.

Also, let me read into register really quick.

**Comment by grantland** [comment]
We actually might have enough for one PR. Let's just clean up the last nits and move the other topics to their own PRs, since they don't really have any impact on any APIs or implementations from this PR. How does that sound?

**Comment by JosephEarl** [comment]
@grantland sure.

I'll update the documentation add `Task.fromCancelled()`. I think it makes sense since in task based continuation it's more efficient to return a cancelled task than throw and it doesn't require adding any features to `CancellationToken`.

**Comment by JosephEarl** [comment]
Like in C# it is the only way to create a cancelled task, so actually I don't think we need it.

**Comment by mingflifb** [comment]
I think adding a CancellableAppLinkResolver makes sense.

**Comment by grantland** [comment]
@JosephEarl: To clarify, we're not adding `Task.fromCancelled(CancellationToken)` since we have `Task.cancelled()`, right?

**Comment by JosephEarl** [comment]
@grantland yes I've left that out. Updated.

File: Bolts/src/bolts/CancellationToken.java [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+package bolts;
+
+import java.util.Locale;
+import java.util.concurrent.CancellationException;
+
+/**
+ * Propagates notification that operations should be canceled.
+ * <p/>
+ * Create an instance of {@code CancellationTokenSource} and pass the token returned from
+ * {@code #getToken()} to the asynchronous operation(s). Call
```
Comment (grantland) [inline Bolts/src/bolts/CancellationToken.java]: {@link CancellationTokenSource#getToken()}

File: Bolts/src/bolts/CancellationToken.java [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+ * {@code CancellationTokenSource#cancel()} to cancel the operations.
+ * <p/>
+ * A {@code CancellationToken} can only be cancelled once - it should not be passed to future operations
+ * once cancelled.
+ *
+ * {@link CancellationTokenSource}
+ *
+ * {@link CancellationTokenSource#getToken()}
+ *
+ * {@link CancellationTokenSource#cancel()}
```
Comment (grantland) [inline Bolts/src/bolts/CancellationToken.java]: @see CancellationTokenSource
    @see CancellationTokenSource#getToken()
    @see CancellationTokenSource#cancel()

File: Bolts/src/bolts/CancellationToken.java [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+ */
+public class CancellationToken {
+
+  private final Object lock = new Object();
+  private boolean cancellationRequested;
+
+  /* package */CancellationToken() {
```
Comment (grantland) [inline Bolts/src/bolts/CancellationToken.java]: space after `/*package */`

File: Bolts/src/bolts/CancellationToken.java [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+   */
+  public void throwIfCancellationRequested() throws CancellationException {
+    synchronized (lock) {
+      if (cancellationRequested) {
+        throw new CancellationException();
+      }
+    }
+  }
+
+  /* package */boolean tryCancel() {
```
Comment (grantland) [inline Bolts/src/bolts/CancellationToken.java]: space after `/*package */`

File: Bolts/src/bolts/CancellationTokenSource.java [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+package bolts;
+
+import java.util.Locale;
+
+/**
+ * Signals to a CancellationToken that it should be canceled. To create a
```
Comment (grantland) [inline Bolts/src/bolts/CancellationTokenSource.java]: {@link CancellationToken}

File: Bolts/src/bolts/CancellationTokenSource.java [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+ * {@code CancellationToken} first create a {@code CancellationTokenSource} then call
```
Comment (grantland) [inline Bolts/src/bolts/CancellationTokenSource.java]: {@link CancellationToken}

File: Bolts/src/bolts/CancellationTokenSource.java [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+ * {@code CancellationTokenSource#getToken()} to retrieve the token for the source.
```
Comment (grantland) [inline Bolts/src/bolts/CancellationTokenSource.java]: {@link #getToken()}

File: Bolts/src/bolts/CancellationTokenSource.java [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * {@link CancellationToken}
+ *
+ * {@link #getToken()}
```
Comment (grantland) [inline Bolts/src/bolts/CancellationTokenSource.java]: @see CancellationToken
    @see #getToken()

File: Bolts/src/bolts/Task.java [hunk 31]
```diff
@@ -196,10 +202,22 @@ public void waitForCompletion() throws InterruptedException {
    * from the callable.
    */
   public static <TResult> Task<TResult> call(final Callable<TResult> callable, Executor executor) {
+    return call(callable, executor, null);
+  }
+
+  /**
+   * Invokes the callable using the given executor, returning a Task to represent the operation.
+   */
+  public static <TResult> Task<TResult> call(final Callable<TResult> callable, Executor executor, final CancellationToken ct) {
```
Comment (grantland) [inline Bolts/src/bolts/Task.java]: line looks a bit long

File: Bolts/src/bolts/Task.java [hunk 32]
```diff
@@ -462,7 +563,16 @@ public Void then(Task<TResult> task) {
...
-    return onSuccess(continuation, IMMEDIATE_EXECUTOR);
+    return onSuccess(continuation, IMMEDIATE_EXECUTOR, null);
+  }
+
+  /**
+   * Runs a continuation when a task completes successfully, forwarding along
+   * {@link java.lang.Exception}s or cancellation.
+   */
+  public <TContinuationResult> Task<TContinuationResult> onSuccess(
+      final Continuation<TResult, TContinuationResult> continuation, final CancellationToken ct) {
```
Comment (grantland) [inline Bolts/src/bolts/Task.java]: line might be over 100 chars

File: README.md [hunk 33]
```diff
@@ -356,6 +356,56 @@ saveAsync(obj1).onSuccessTask(new Continuation<ParseObject, Task<ParseObject>>()
...
+        }
+        result++;
+      }
+      tcs.setResult(result);
+    }
+  }.start();
+
+  return tcs.getTask();
+}
+
```
Comment (grantland) [inline README.md]: nit: remove newline

**Comment by grantland** [comment]
Implementation looks good, just a few doc nits. I'll also assume you're adding `CancellationToken.register()` in another PR, since it's not in this one.

Thanks!

**Comment by JosephEarl** [comment]
@grantland Updated docs as per your comments.

Yes I will move `CancellationToken.register()` to another PR - I'd rather get this approved and submit a matching PR for iOS first :)

**Comment by grantland** [comment]
Looks great, thanks! I'll look forward to the rest of it :)

**Pull request merged**